odoo/http.py

```diff
@@ -336,10 +336,13 @@ def _call_function(self, *args, **kwargs):
         def checked_call(___dbname, *a, **kw):
             # The decorator can call us more than once if there is an database error. In this
             # case, the request cursor is unusable. Rollback transaction to create a new one.
-            if self._cr:
+            try:
+                result = self.endpoint(*a, **kw)
+            except psycopg2.InternalError as e:
+                if not self._cr or not e.pgcode == psycopg2.errorcodes.IN_FAILED_SQL_TRANSACTION:
```

Comment:
No need to check the presence of `self._cr`, will always be true.
I would also rollback in all cases.